### PR TITLE
Created AVL tree in C++( issue  #2112)

### LINKED ISSUE #2112: AVL tree
- [x] C
- [x] C++ - @Manish-cloud
- [ ] C# - @KanakalathaVemuru
- [ ] CoffeeScript
- [x] Dart - @Rukmini-Meda
- [ ] Go
- [x] Java
- [ ] JS - @satvikchachra
- [ ] Kotlin
- [ ] Php - @chandrikadeb7
- [x] Python
- [x] Readme
- [ ] Ruby
- [ ] Rust
- [ ] TypeScript

### PR DESCRIPTION
**Pull request by aryangulati**
Fixes #2112

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Create AVL tree 
- 

#### Languages Used:
- C++(cpp) 
- 

#### Files Added:
- AVL_Tree/AVL_Tree.cpp
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by jainaman224** [comment]
Refer to issue number.

File: AVL_Tree/AVL_Tree.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
+#include<bits/stdc++.h>
+#define pow2(n) (1 << (n))
+using namespace std;
+struct avl {
+   int d;
+   struct avl *l;
+   struct avl *r;
+}*r;
+class avl_tree {
```
Comment (jainaman224) [inline AVL_Tree/AVL_Tree.cpp]: Add a new line above blocks.

File: AVL_Tree/AVL_Tree.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+   public:
```
Comment (jainaman224) [inline AVL_Tree/AVL_Tree.cpp]: Use 4 space indentation.

File: AVL_Tree/AVL_Tree.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+   if (t == NULL)
+      return;
+      cout << t->d << " ";
+      preorder(t->l);
+      preorder(t->r);
+}
+void avl_tree::postorder(avl *t) {
+   if (t == NULL)
+      return;
+      postorder(t ->l);
```
Comment (jainaman224) [inline AVL_Tree/AVL_Tree.cpp]: Incorrect indentation.

**Comment by jainaman224** [comment]
@aryangulati Are you working on it?

**Comment by somya-kapoor** [comment]
@aryangulati Are you working on it?

**Comment by aryangulati** [comment]
Yed

**Comment by aryangulati** [comment]
please @jainaman224 review it once!!

File: AVL_Tree/AVL_Tree.cpp [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: incorrect indentation here

File: AVL_Tree/AVL_Tree.cpp [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+            exit(1);
+         break;
+         default:
+            cout << "Wrong Choice" << endl;
+      }
+   }
+   return 0;
+}
+
+
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: remove these extra lines

File: AVL_Tree/AVL_Tree.cpp [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+      preorder(t->r);
+}
+void avl_tree::postorder(avl *t) {
+   if (t == NULL)
+      return;
+      postorder(t ->l);
+      postorder(t ->r);
+      cout << t->d << " ";
+}
+int main() {
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: add a new line above `int main()`

File: AVL_Tree/AVL_Tree.cpp [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+   }
+}
+void avl_tree::inorder(avl *t) {
+   if (t == NULL)
+      return;
+      inorder(t->l);
+      cout << t->d << " ";
+      inorder(t->r);
+}
+void avl_tree::preorder(avl *t) {
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: add a new line above each function

File: AVL_Tree/AVL_Tree.cpp [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+   return rr_rotat(parent);
+}
+avl *avl_tree::balance(avl *t) {
+   int bal_factor = difference(t);
+   if (bal_factor > 1) {
+      if (difference(t->l) > 0)
+         t = ll_rotat(t);
+      else
+         t = lr_rotat(t);
+   } else if (bal_factor < -1) {
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: `else if` should be in new line

File: AVL_Tree/AVL_Tree.cpp [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+   return t;
+}
+avl *avl_tree::insert(avl *r, int v) {
+   if (r == NULL) {
+      r = new avl;
+      r->d = v;
+      r->l = NULL;
+      r->r = NULL;
+      return r;
+   } else if (v< r->d) {
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: same here

**Comment by aryangulati** [comment]
Changes made please review it!!

File: AVL_Tree/AVL_Tree.cpp [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+
+avl *avl_tree::balance(avl *t) {
+   int bal_factor = difference(t);
+   if (bal_factor > 1) {
+      if (difference(t->l) > 0)
+         t = ll_rotat(t);
+      else
+         t = lr_rotat(t);
+   }
+    else if (bal_factor < -1) {
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: use 4 space indentation.

File: AVL_Tree/AVL_Tree.cpp [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+
+avl *avl_tree::insert(avl *r, int v) {
+   if (r == NULL) {
+      r = new avl;
+      r->d = v;
+      r->l = NULL;
+      r->r = NULL;
+      return r;
+   } 
+    else if (v< r->d) {
```
Comment (raghav-dalmia) [inline AVL_Tree/AVL_Tree.cpp]: same here

**Comment by raghav-dalmia** [comment]
Also, squash your commits. Refer [this](https://github.com/wprig/wprig/wiki/How-to-squash-commits).

**Comment by aryangulati** [comment]
done

**Comment by aryangulati** [comment]
@raghav-dalmia please review it changes and squash is done

**Comment by somya-kapoor** [comment]
@aryangulati Was this issue assigned to you?

**Comment by aryangulati** [comment]
Yes @somya-kapoor please review it!!!

**Comment by somya-kapoor** [comment]
But the issue #2112 C++ is assigned to someone else.

**Comment by aryangulati** [comment]
It's assigned to me and also too you are not reviewing pls contact @raghav-dalmia and pls pls pls review tell me further changes so atleast my first pr got merged!! @somya-kapoor

**Comment by somya-kapoor** [comment]
@aryangulati Who assigned it to you? Because it is already assigned to someone else.

**Comment by aryangulati** [comment]
I don't know bcz it's been  a month
Can u do me a favor that please unsigned all the issue so that i can make
my first contribution successfully and have a new fresh start as I don't
know!!

**Comment by somya-kapoor** [comment]
It is already assigned to someone, so I can't merge your PR. Also, I checked the issue and it was not assigned to you.

**Pull request closed without merging**